pandas/io/parquet.py

```diff
@@ -0,0 +1,140 @@
+        raise ValueError("engine must be one of 'pyarrow', 'fastparquet'")
+
+
+def to_parquet(df, path, engine, compression=None):
+    """
+    Write a DataFrame to the pyarrow
+
+    Parameters
+    ----------
+    df : DataFrame
+    path : string
+        File path
+    engine : parquet engine
+        supported are {'pyarrow', 'fastparquet'}
+    compression : str, optional
+        compression method, includes {'gzip', 'snappy', 'brotli'}
+    """
+
+    _validate_engine(engine)
+
+    if not isinstance(df, DataFrame):
+        raise ValueError("to_parquet only support IO with DataFrames")
+
+    valid_types = {'string', 'unicode'}
+
+    # validate index
+    # --------------
+
+    # validate that we have only a default index
+    # raise on anything else as we don't serialize the index
+
+    if not isinstance(df.index, Int64Index):
+        raise ValueError("parquet does not serializing {} "
+                         "for the index; you can .reset_index()"
+                         "to make the index into column(s)".format(
+                             type(df.index)))
+
+    if not df.index.equals(RangeIndex.from_range(range(len(df)))):
+        raise ValueError("parquet does not serializing a non-default index "
+                         "for the index; you can .reset_index()"
+                         "to make the index into column(s)")
+
+    if df.index.name is not None:
+        raise ValueError("parquet does not serialize index meta-data on a "
+                         "default index")
+
+    # validate columns
+    # ----------------
+
+    # must have value column names (strings only)
+    if df.columns.inferred_type not in valid_types:
+        raise ValueError("parquet must have string column names")
+
+    if engine == 'pyarrow':
+        pyarrow = _try_import_pyarrow()
+        from pyarrow import parquet as pq
+
+        table = pyarrow.Table.from_pandas(df)
+        pq.write_table(table, path, compression=compression)
+
+    elif engine == 'fastparquet':
+        fastparquet = _try_import_fastparquet()
+
+        # thriftpy/protocol/compact.py:339:
+        # DeprecationWarning: tostring() is deprecated.
+        # Use tobytes() instead.
+        with catch_warnings(record=True):
+            fastparquet.write(path, df, compression=compression)
+
+
+def read_parquet(path, engine):
+    """
+    Load a parquet object from the file path
+
+    .. versionadded 0.20.0
+
+    Parameters
+    ----------
+    path : string
+        File path
+    engine : parquet engine
+        supported are {'pyarrow', 'fastparquet'}
+
+    Returns
+    -------
+    type of object stored in file
+
+    """
+
+    _validate_engine(engine)
+
+    if engine == 'pyarrow':
+        pyarrow = _try_import_pyarrow()
+        return pyarrow.parquet.read_table(path).to_pandas()
+
+    elif engine == 'fastparquet':
+        fastparquet = _try_import_fastparquet()
+        pf = fastparquet.ParquetFile(path)
+        return pf.to_pandas()
```

Comment:
you could encapsulate the details for both implementations in a `FastparquetImpl` and `PyarrowImpl` which have the same API